# IBDHap-detector
 
IBDHap-detector is pipeline of detecting IBD or Haplotype segements based on phasing genotypes.

# Installation
Requires HMM, pbapply, parallel, randomForest, MASS and smotefamily packages in R

# data preparation
 You need to prepare correct.haps.filtered.formated.txt, formated_w221_snp.filtered.txt and genetic maps first.
 
 correct.haps.filtered.formated.txt gives the following format:
 ```
 m_chr1_40	1_57438086	G	A	1	74423
 m_chr1_40	1_57438105	G	A	1	74442
 m_chr1_40	1_57438151	C	T	1	74488
 m_chr1_40	1_57438276	C	A	1	74613
 ```
 The first column is bin id, the second column show the chromosome and position of the SNP in DASZ genome, which are seprated by the underline. The third and fourth columns indicate the two phased SNPs. The last two columns are chromosome and position of the SNP in Fudingdabai genome.
 
 formated_w221_snp.filtered.txt gives the following format:
 ```
 X3W1.1	X3W1.10	X3W1.12	X3W1.13	X3W1.14	X3W1.15	X3W1.16
 1_107784	AG	A	A	A	A	NA	A
 1_108003	GA	G	G	G	G	GA	G
 ```
 Rownames indicate the chromosome and position of the SNP in DASZ genome, which is same as the second column the  correct.haps.filtered.formated.txt. Colnames show the sample IDs. Missing genotypes are showing as NA.
 
 The example of genetic maps are in the directory "linkage_group", which is generated by MSTmap software.
 
# Running
These scripts could be used to reproduce the results in our papers. If you want to use them to analyze your own data, you may need to transfer your data to the suitable format for this pipeline. 

## Step 1:
```
Rscript 00Pre_stat_phasing_fudingHaps.R
```
This step compare the phased haploypes in correct.haps.filtered.formated.txt to the genotypes of each sample in formated_w221_snp.filtered.txt and generate 00res_fuding_haps_all_for_plot.Rdata and 00res_fuding_haps_all_for_plot.txt for the next steps.

## Step 2:
```
Rscript 01plot_haps_foreach_accessions.R
```
This step produce the example IBD plot which could help you to select samples for the training sets of random forest model.

## Step 3: 
Change the dir to discriminant_analysis and run:
```
Rscript 00phasing_fudingHaps_generate_data_for_discriminant_analysis.R
```
This step will generate training_samples_stat_info.txt and you can combine the IBD plot in Step 2 to assign the IBD segments and save the results in test_training.txt

## Step 4:
```
Rscript 01discriminant_anaysis
```
Generate RandomForest model

## Step 5:
Back to the main directory and run:
```
Rscript 02phasing_fudingHaps_from_Rrandom_forest
```
This step predict IBD segments in tea population based on random forest model.

## Step 6:
```
Rscript 03calculaing_IBD.R
Rscript 04plot_pred_haps.R
```
Calculating IBD proportion and plot the IBD detection results

# Citation:
Zhang, Weiyi; Luo, Cheng; Scossa, Federico; Zhang, Qinghua; Usadel, Bjoern; Fernie, Alisd air R.; Mei, Hanwei; Wen, Weiwei; A phased genome based on single sperm sequencing reveals crossover pattern and complex relatedness in tea plants, Plant J. 2021, 105(1): 197-208
